File: 2835 - A Round Peg in a Ground Hole/holes.cpp
#include <iostream>
#include <fstream>

using namespace std;

int nVertices;

struct Point {
  double x;
  double y;
};

ostream& operator<< (ostream& out, Point p)
{
  out << "(" << p.x << ", " << p.y << ")";
  return out;
}


Point* vertices = 0; 
double pegSize;
Point pegPosition;

bool debugging = false;



double edgeEval (int edgeNum, Point p)
{
  Point p1 = vertices[edgeNum];
  Point p2 = vertices[(edgeNum + 1) % nVertices];

  double r = (p1.y - p2.y)*(p.x - p2.x) - (p.y - p2.y)*(p1.x - p2.x);
  return r;
}


Point center()
{
  Point p = {0.0, 0.0};
  for (int i = 0; i < nVertices; ++i)
    {
      p.x += vertices[i].x;
      p.y += vertices[i].y;
    }
  p.x /= (double)nVertices;
  p.y /= (double)nVertices;
  return p;
}


bool testForConvexity()
{
  Point c = center();
  for (int v = 0; v < nVertices; ++v)
    for (int e = 0; e < nVertices; ++e)
      {
	double rc = edgeEval(e, c);
	double rv = edgeEval(e, vertices[v]);
	if (rc * rv < 0.0)
	  return false;
      }
  return true;
}

double distSqr(Point p1, Point p2)
{
  return ((p2.x - p1.x)*(p2.x - p1.x) + (p2.y - p1.y)*(p2.y - p1.y));
}

double dist(Point p1, Point p2)
{
  return sqrt(distSqr(p1, p2));
}


Point closestPoint (int edgeNum, Point p3)
{
  Point p1 = vertices[edgeNum];
  Point p2 = vertices[(edgeNum+1) % nVertices];
  double u = (p3.x - p1.x) * (p2.x - p1.x) + (p3.y - p1.y) * (p2.y - p1.y);
  u = u / distSqr(p1, p2);
  if (debugging)  cerr << "u = " << u << endl;
  if (u < 0.0)
    return p1;
  else if (u > 1.0)
    return p2;
  else
    {
      Point p;
      p.x = p1.x + u * (p2.x - p1.x);
      p.y = p1.y + u * (p2.y - p1.y);
      return p;
    }
}



bool testPegPosition()
{
  double peg2 = (pegSize*pegSize);
  for (int e = 0; e < nVertices; ++e)
    {
      Point p = closestPoint(e, pegPosition);
      double d = distSqr(pegPosition,p);
      if (d < peg2)
	return false;
    }
  Point c = center();
  for (int e = 0; e < nVertices; ++e)
    {
      double rc = edgeEval(e, c);
      double rp = edgeEval(e, pegPosition);
      if (rc * rp < 0.0)
	return false;
    }
  return true;
}
  



void holes (istream& in)
{
  in >> nVertices;
  while (nVertices > 2)
    {
      in >> pegSize >> pegPosition.x >> pegPosition.y;
      if (debugging)
	cerr << nVertices << " "
	     << pegSize << " "
	     << pegPosition << endl;
      delete [] vertices;
      vertices = new Point[nVertices];

      for (int i = 0; i < nVertices; ++i)
	{
	  in >> vertices[i].x >> vertices[i].y;
	  if (debugging)
	    cerr << vertices[i] << endl;
	}


      if (testForConvexity())
	{
	  if (testPegPosition())
	    cout << "PEG WILL FIT" << endl;
	  else
	    cout << "PEG WILL NOT FIT" << endl;
	}
      else
	cout << "HOLE IS ILL-FORMED" << endl;

      in >> nVertices;
    }
}


int main (int argc, char** argv)
{
  if (argc > 1)
    {
      ifstream in (argv[1]);
      debugging = true;
      holes(in);
    }
  else
    holes(cin);

  return 0;
}

File: 2835 - A Round Peg in a Ground Hole/holes.java
import java.io.*;
import java.util.*;

class coord {
  double x, y;

  public String toString(){
    return "("+x+","+y+")";
  } 
}

public class holes {

    public static boolean onLeftSide(coord point, coord linea, coord lineb){
	double dy = lineb.y - linea.y;
	double dx = lineb.x - linea.x;
	boolean onLeft;
	
	if(dx == 0){
	    onLeft = point.x < linea.x;
	    if(dy < 0)
		onLeft = !onLeft;
	}
	else{
	    double slope = dy/dx;
	    
	    double lineY = slope*(point.x - linea.x)+linea.y;
	    
	    onLeft = point.y > lineY;

	    if(dx<0 )
		onLeft = !onLeft;

	}


	System.err.println("point "+point+" is on left of "+linea+"-"+lineb+
			   ": "+onLeft);

	return onLeft;
    }


  public static double distPointLine(coord point, coord line, coord slopeV){

    if(slopeV.x == 0)
      return Math.abs(point.x - line.x);
    
    double m = slopeV.y/slopeV.x;
    double a = -m;
    double b = 1;
    double c = m*line.x-line.y;

    return Math.abs(a*point.x+b*point.y+c)/Math.sqrt(a*a+b*b);
  }

  public static coord vector(coord a, coord b){
    coord vector = new coord();
    vector.x = b.x-a.x;
    vector.y = b.y-a.y;
    return vector;
  }
    
  public static double length(double x1, double y1, double x2, double y2){
    return Math.sqrt((x1-x2)*(x1-x2)+(y1-y2)*(y1-y2));
  }

  public static double length(coord vector){
    return Math.sqrt(vector.x*vector.x + vector.y*vector.y);
  }

  public static void main(String[] argv)
    throws Exception {
    BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

    SHAPE:
    while (true){
      coord[] vertices;

      coord peg = new coord();

      StringTokenizer t = new StringTokenizer(in.readLine());
      int Nvert = Integer.parseInt(t.nextToken());
      if(Nvert < 3)
	break;
      double pegRadius = Double.parseDouble(t.nextToken());
      peg.x = Double.parseDouble(t.nextToken());
      peg.y = Double.parseDouble(t.nextToken());

      vertices = new coord[Nvert];

      for(int i=0;i<Nvert;i++){
	vertices[i] = new coord();
	t = new StringTokenizer(in.readLine());
	vertices[i].x = Double.parseDouble(t.nextToken());
	vertices[i].y = Double.parseDouble(t.nextToken());
      }

      double angleSum = 0 ;
      double thisAngle;
      for(int i=0;i<Nvert;i++){
	coord va = vector(vertices[i], vertices[(i+1)%Nvert]);
	coord vb = vector(vertices[(i+1)%Nvert],vertices[(i+2)%Nvert]);
	thisAngle= Math.acos((va.x*vb.x +
			      (va.y*vb.y))/
			     (length(vb)*length(va)));
	thisAngle = Math.toDegrees(thisAngle);
	angleSum+=thisAngle;
	System.err.println("got indiv angle "+thisAngle);
      }
      System.err.println("total "+angleSum);
    
      if(angleSum > 360.01){
	System.out.println("HOLE IS ILL-FORMED");
	continue;
      }

      for(int i=0;i<Nvert;i++){
	double dist = distPointLine(peg, vertices[i], vector(vertices[i],vertices[(i+1)%Nvert]));
	System.err.println("distance of line from "+vertices[i]+" at slope "+
			   vector(vertices[i],vertices[(i+1)%Nvert])+" from peg "+peg+
			   " = "+dist);
	if (dist < pegRadius){
	  System.out.println("PEG WILL NOT FIT");
	  continue SHAPE;
	}

      }

      

      boolean firstLeft = onLeftSide(peg, vertices[Nvert-1],vertices[0]);
      boolean allSame = true;

      for(int i=0;i<Nvert-1;i++){
	  if(onLeftSide(peg, vertices[i],vertices[i+1])!=firstLeft){
	      System.out.println("PEG WILL NOT FIT");
	      continue SHAPE;
	  }
      }


      System.out.println("PEG WILL FIT");
    }
  }
}

